### PR TITLE
Alerting: Move fake rule store to the test utilities package

### PR DESCRIPTION
**Pull request by alexweav**
**What this PR does / why we need it**:

The previous PR made the rule store interface no longer bound to its implementation: https://github.com/grafana/grafana/pull/55968

This means, we can now have alternative implementations of rule store, living in separate packages! One such implementation is our `FakeRuleStore`.

This PR moves `FakeRuleStore` to a test utilities package, so it can be easily used in a lot of places. Due to our interface splitting work from earlier, it can take very few dependencies (only `model`), so it's easy to use from anywhere you need. Packages no longer need to depend directly on `store` for tests, which breaks many package dependencies.

This is based on a suggestion from @yuri-tceretian - thanks! It was not possible until the we broke the interface dependency in the previous PR.

**Which issue(s) this PR fixes**:

Contributes to https://github.com/grafana/grafana/issues/55770

**Special notes for your reviewer**:

### CODE REVIEW

**Review by joeblubaugh — COMMENTED** [review]
This change kind of re-binds all the test code to an implicit interface, shared between two packages. You have to change the `fakes.FakeRuleStore` (maybe name it `fakes.RuleStore`?) and `store.DBStore` in unison now, rather than keep package-specific test code in those packages. This can probably get even simpler by implementing smaller fakes within each package instead of sharing a type that implements multiple interfaces.

Right now you're also binding `ngalert` to an `api` package interface, when the minimal required interface is a single function.

**Comment by alexweav** [comment]
> This can probably get even simpler by implementing smaller fakes within each package instead of sharing a type that implements multiple interfaces.

100% agree. I've already done this in a few cases, like in the `schedule` package. Some packages used very narrow slices of the RuleStore methods, when this happened I just wrote a separate fake in that package just for those tests, this was super nice and the tests are a lot more clear in general. You have an interface, fakes, tests, and consumers all in one package, everything is self contained and easy to find.

The only tests that still actually depend on this fake are _very_ complex, and you can also see vice versa, that the fake itself is very complex to support these tests. This fake and the last 2 packages are a bit of a tangled mass that can't be so quickly refactored.

>  implicit interface

In practice, it's not quite so implicit.. The only packages that now depend on this use nearly all of the RuleStore methods. I've extracted functionality from this store where I can, but it's now down to just that chunk remaining.

This PR at a minimum decouples those packages from `store` and eliminates a lot of the package dependencies causing import problems. Simplifying the rule-related tests themselves is another area of debt, but I'm trying to keep these refactors relatively scoped. WDYT @joeblubaugh

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (joeblubaugh) [inline pkg/services/ngalert/tests/fakes/rules.go:19]: Maybe for a follow-up PR, but is it worth just using the store.DBStore with sqlite3 when tests are complex enough to need this type? There's enough logic in here that it's an independent in-memory implementation rather than a deterministic fake, so maybe these complex tests are better expressed as integration tests between the DBStore and the calling package.

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (joeblubaugh) [inline pkg/services/ngalert/tests/fakes/rules.go:25]: I grepped through, and we only verify RecordedOps in a couple of places. I don't think it's a good system - It would be better to just verify that the state of the database is as expected than cart around this unwieldy RecordedOps setup, switching on the arg type and counting against expectations. What you should see is that a rule you wanted is no longer available, not the # of times you called `Delete`.

Let's file a ticket to nix this thing.

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (joeblubaugh) [inline pkg/services/ngalert/tests/fakes/rules.go:24]: Let's file a ticket to delete this, too. The tests that use it would be better served by a Fake whose implementation of `GetAlertRulesBy...` could be set to return an error than a Hook that uses type assertions.

**Review by joeblubaugh — APPROVED** [review]
LGTM. I hope we will continue to dismantle this system - a "fake" this complex gives me the heebie jeebies.

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/tests/fakes/rules.go:25]: Recorded Ops, as well as Hook below, are the sign that we need just a mock instead of fake. 

> to just verify that the state of the database is as expected than cart around this unwieldy RecordedOps setup,

Then to verify a state the fake would have to simulate the database. This is even more code in the fake and logic that also needs to be tested (tests for fakes). I would rather replace it with mocks.

**Review by yuri-tceretian — APPROVED** [review]
The reason I suggested keeping a single fake rather than implementing many small ones is that it will cause unnecessary data duplication. The end goal in my opinion should be either replacement to mocks or changing the production code and then tests so they could be tested without having to use RecordedOps and Hooks.

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (alexweav) [reply pkg/services/ngalert/tests/fakes/rules.go:25]: My first instinct with this was actually to just go ahead and refactor tests to use a mock rather than move the store.

I actually got semi-far with this. A lot of tests pass with the mock. Some edge cases were not so easy, though. Some of our tests rely heavily on the filtering logic in the fake (which means they're testing the fake itself, not the prod code lol). You'd `PutRule`, and then `ListRules` would only respond with a subset of the expected rules because of the filtering.

These were pretty difficult to pull apart, so I put that branch on hold for now.

Just noting that I agree with the two of you - this was the thing I tried first.

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (alexweav) [reply pkg/services/ngalert/tests/fakes/rules.go:24]: This one is simple enough, I added it as a task on the current refactor.

File: pkg/services/ngalert/tests/fakes/rules.go
Comment (alexweav) [reply pkg/services/ngalert/tests/fakes/rules.go:19]: This is what we did for Provisioning, it's definitely nicer. We didn't want to add further dependencies on this thing. It's a pain to maintain. Some other places use a mock instead of a fake, that's also been a huge improvement. I agree that this fake has gotten out of hand, and I've pared it down where I can.

I'm trying to limit the scope of this refactor a bit, it feels we're falling into the syndrome of finding something new that's broken every time we go to fix something. That's why I didn't spend too much time pulling apart particularly tough blobs like this. We removed a number of consumers of this fake already as part of this effort, that's significant progress.

The purpose of refactor is to unblock the state history work, and with this PR it seems we've reached that point. I'd like to wrap up any loose ends, and start filing remaining issues into the backlog rather than trying to fix the world here. I'll create an issue noting this, but I'm not sure if I'll be able to come back and actually rework these tests soon.

Comment (alexweav) [reply pkg/services/ngalert/tests/fakes/rules.go:19]: I like it!

**Pull request merged**
